lib/server.js

```diff
@@ -0,0 +1,205 @@
+import http from 'node:http'
+import net from 'node:net'
+import {EventEmitter} from 'node:events'
+import crypto from 'node:crypto'
+import url from 'node:url'
+
+import {Headers} from 'undici'
+
+import {SafeHttpClient, HttpError} from './safe-http-client.js'
+import {
+  DEFAULT_HEADERS,
+  REQUEST_HEADERS,
+  RESPONSE_HEADERS
+} from './constants.js'
+
+/**
+ * @typedef {Object} Config
+ * @property {string} HMACKey
+ * @property {string} serverName Server name used in Headers and Via checks
+ * @property {number|undefined} [maxSize] The max size in bytes per resource to download. Sends 500 if exceeded.
+ */
+
+/**
+ * @param {string} hex
+ */
+function hexDecode(hex) {
+  const bytes = []
+  for (let i = 0; i < hex.length; i += 2) {
+    bytes.push(parseInt(hex.substr(i, 2), 16))
+  }
+  return Buffer.from(bytes)
+}
+
+/**
+ * Filter incoming headers based on allowed headers.
+ * We always want to send Pascal-Case headers so we use the key from constants
+ * with the value from the incoming header.
+ * @param {Map<string, boolean>} allowedHeaders
+ * @returns {(incomingHeaders: Headers) => Record<string, string>}
+ * */
+function filterHeaders(allowedHeaders) {
+  return function (incomingHeaders) {
+    /** @type {Record<string, string>} */
+    const filteredHeaders = {}
+
+    for (const [key, bool] of allowedHeaders.entries()) {
+      if (incomingHeaders.has(key.toLowerCase()) && bool) {
+        // @ts-ignore
+        filteredHeaders[key] = incomingHeaders.get(key.toLowerCase())
+      }
+    }
+
+    return filteredHeaders
+  }
+}
+
+/*
+ * Node.js HTTP server to proxy insecure content via HTTPS.
+ */
+export class Server extends EventEmitter {
+  /**
+   * @param {Config} config
+   */
+  constructor(config) {
+    super()
+    if (!config.HMACKey) {
+      throw new Error('HMACKey is required')
+    }
+    this.config = config
+    this.config.serverName ??= 'camo'
```

Comment:
```suggestion
    this.config = {...config, serverName: config?.serverName || 'camo'}
```

Do not mutate the given object